### PR TITLE
added form validation for appointment creation

### LINKED ISSUE #276: Member: Do error-checking on form and give error messages for Schedule an Appointment
Currently on pages like https://circulate-staging.herokuapp.com/appointments/new?
1. members can schedule an appointment without selecting any items, and
2. member can submit without selecting a date, which gives a 404-like page.

These are both undesirable. To fix this:

For (1), check the form to make sure at least one item is selected before submitting. If no item is selected, display an error message, "Please select an item to pick-up or return for your appointment."

For (2), check the form to make sure a date and time is selected from the pulldown. If no date is selected, display an error message, "Please select a date and time for this appointment."

There may be other error messages we haven't thought of that should be incorporated as well. Feel free to catch those cases and create error messages for them as well.

### PR DESCRIPTION
**Pull request by renugasaraswathy**
Fixes #276 

Please find the screenshots with the error messages displayed.

**Screen 1**: Appointment Form

Screenshot 2020-10-06 at 7 27 41 AM

**Screen 2**: Form is submitted without selecting item and date. Shows respective error messages

Screenshot 2020-10-06 at 7 27 51 AM

**Screen 3**: Form is submitted by selecting item and not selecting date. Shows error message only for date.

Screenshot 2020-10-06 at 7 28 05 AM

**Screen 4**: Form is submitted by selecting date and not selecting item. Shows error message only for item.

Screenshot 2020-10-06 at 7 28 43 AM

**Screen 5**: Form is submitted by selecting date and selecting item. Appointment is saved successfully.

Screenshot 2020-10-06 at 7 29 05 AM

### CODE REVIEW

**Comment by rkim00** [comment]
Looks great! Thanks for all the screen shots, and thanks for your work!

**Comment by renugasaraswathy** [comment]
Hi @rkim00, Thank you for your feedback. Could you please get it reviewed by the tech leads. 

Thanks

**Review by mdworken — CHANGES_REQUESTED** [review]
Looks like a really excellent start! I left a couple comments within the PR on some code that I'm not sure still needs to be in here.

My only other comment is that we should probably use the built in rails presence validators for these, unless there's a specific reason not to.

File: app/models/appointment.rb [hunk 1]
```diff
@@ -5,14 +5,28 @@ class Appointment < ApplicationRecord
   has_many :loans, through: :appointment_loans
   belongs_to :member
 
-  validates :starts_at, :ends_at, presence: true
-
-  validate :ends_at_later_than_starts_at
+  validate :ends_at_later_than_starts_at, :item_present, :date_present
+  attr_accessor :time_range_string
```
[the remark this answers is not in the document]
Comment (renugasaraswathy) [reply app/models/appointment.rb:9]: Yes right, will remove it

File: app/models/appointment.rb [hunk 2]
```diff
@@ -5,14 +5,28 @@ class Appointment < ApplicationRecord
...
 
   scope :upcoming, -> { where("starts_at > ?", Time.zone.now).order(:starts_at) }
 
   private
 
+
+  def item_present
+    puts "inside item_present"
+    puts holds.inspect
```
[the remark this answers is not in the document]
Comment (renugasaraswathy) [reply app/models/appointment.rb:18]: Yes. Missed removing the puts statements, will remove them.

File: app/views/appointments/_form.html.erb [hunk 3]
```diff
@@ -17,7 +24,7 @@
       <tbody>
         <% loans.each do |loan| %>
           <tr>
-            <td><%= check_box_tag("appointment[loan_ids][]", loan.id) %></td>
+            <td><%= check_box_tag("appointment[loan_ids][]", loan.id, (@appointment.loans.include?(loan))) %></td>
```
[the remark this answers is not in the document]
Comment (renugasaraswathy) [reply app/views/appointments/_form.html.erb:27]: Forms wont be pre-filled when user sees the page for the first time. 

It is pre-filled only when the form is not saved and there are errors. It is a good user experience if it is pre-filled and not ask the user to fill the entire form when a mistake is made.

File: app/controllers/appointments_controller.rb [hunk 4]
```diff
@@ -25,6 +27,8 @@ def create
     if @appointment.save
       redirect_to appointments_path, notice: "Appointment was successfully created."
     else
+      @holds = Hold.active.where(member: current_user.member)
+      @loans = current_user.member.loans.includes(:item).checked_out
```
[the remark this answers is not in the document]
Comment (renugasaraswathy) [reply app/controllers/appointments_controller.rb:31]: These variables are needed to render the check boxes when form has errors.

When there are errors, the flow would be

Controller new -> new.html.erb -> submit form -> controller create method -> if errors -> renders new.html.erb

So whatever variables used in new.html.erb, to be set values in 'def create' if there are errors.

**Comment by renugasaraswathy** [comment]
Hi @mdworken, thank you so much for your time and feedback. I have given my replies in the PR. 

Thanks

**Comment by renugasaraswathy** [comment]
@jim @johnksawers could you review this PR. Thanks.

**Comment by jim** [comment]
@renugasaraswathy Thanks for this contribution! I will review this shortly.

**Review by jim — APPROVED** [review]
Tested and works as described in the story! Thanks very much for the contribution.

**Comment by rkim00** [comment]
@jim These changes don't seem to be showing up in staging. Is there any reason they might have gotten lost?

**Comment by rkim00** [comment]
This now appears to be working in staging.

**Pull request merged**